Add subscription item table to our customer notification emails and introduce before/after hooks for further customisations

mattallan opened the pull request:
Fixes #752

## Description

<!--
Write a brief summary about this PR. 
- Why is this change needed? 
- What does this change do? 
- Were there other solutions you considered? 
- Why did you choose to pursue this solution? 
- Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos as appropriate.
-->

We received the following feedback related to ur new [Subscriptions notification feature](https://woocommerce.com/document/subscriptions/subscriptions-notifications/) we released in 6.9.0: 
> the information on the "Subscription information" email template might not be enough for the customer to understand why is he getting this email.

> there's absolutely no information about the subscription items. Yes, they can click the subscription ID, log in, and see its details, but we have a serious risk here that the customer just ignores this email, as they see no identifiable information

The author opened a PR (https://github.com/Automattic/woocommerce-subscriptions-core/pull/753) to address these issues, however, while I was reviewing the PR, there were a few things I wanted to change so I figured it was easier to open a new PR instead of working from their fork.

Key changes in this PR:
- 0e10ca601537b9573efb2e99e9034bf59324f523 - Adds the subscription item table to all of our customer notification emails.
- 9f2f7fbfe2158a9f0a665c805f6478176bb437e3 - Adds new hooks at the start and end of the `subscription-info.php` template for further customisations
- 9aa0ad6d219d9c08eb49f0342c1b7792295a923a - Removes the "Here are the details" paragraphs as it felt unnecessary.

| Before | After |
|--------|--------|
| ![image](https://github.com/user-attachments/assets/202a1088-fda1-483b-ab50-2e0ad1e27bf2) | ![image](https://github.com/user-attachments/assets/6e82c4ca-871b-49ae-8843-bdfdfa6442b1) | 

> [!NOTE]
> When I started looking more closely at this issue, I noticed there are a lot of inconsistencies across many of our subscription emails. For example:
> - our Cancelled email doesn't have the `"Subscriptions #123` header between the two tables, but our Expired and Suspended emails do (see images below)
> - these emails show a different subscription info table where the recurring price is the second column, vs our customer notification emails have it in the 4th column
>
> 🙈 I had to stop myself from turning PR this into a massive refactor... I do think we should tackle this at some stage, but maybe it can wait for the Woo core email improvements that are also in the pipeline.

| Cancelled email | Expired email |
|--------|--------|
| ![image](https://github.com/user-attachments/assets/0536ff5e-9e4d-4a66-8b1b-a243a89d57a8) | ![image](https://github.com/user-attachments/assets/cfda40f6-e19b-424f-b022-a0296973d83b) | 

## How to test this PR

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
Use the testing instructions from the linked issue as a starting point.
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

With WooCommerce 9.6 installed, you are to preview emails within **WooCommerce > Settings > Emails** which makes testing this feature a little easier. That said, it uses dummy data and doesn't use the plain text templates properly, therefore testing this PR requires:
1. Installing and activate the [WP Mail Logging](https://wordpress.org/plugins/wp-mail-logging/) plugin for ease with testing emails locally.
2. Adding `define( 'WP_ENVIRONMENT_TYPE', 'production' );` to your `wp-config.php` file. This is needed because notification emails are not sent on development/test sites.

### 1. Testing email template changes

1. Go to **WooCommerce > Settings > Subscriptions** and enable the "Customer Notification" feature (set any reminder time)
2. Go to **WooCommerce > Settings > Emails**
3. Enable the "Customer Notification: Automatic renewal notice" email and make sure it's set to HTML format.
4. Enable the "Customer Notification: Manual renewal notice" email and set this one to **plain text** format
![image](https://github.com/user-attachments/assets/b8f78149-c1df-4ff5-b211-138ee2e1bea7)
5. Visit the WP Admin Edit subscription page for a subscription purchased with Stripe (or any other automatic payment method).
6. From the subscription actions, run the "Send upcoming renewal notification" action.
![image](https://github.com/user-attachments/assets/2c4b56fb-abb5-4d3b-927f-10ab0a5b3e52)
8. View the email and confirm you see an email with both the subscription info table and subscription items table:
![image](https://github.com/user-attachments/assets/259ec9a1-08be-4ace-99fa-a7b91e507806)
9. Go back to Admin Edit Subscription and switch the subscription from Stripe to manual renewal:
![image](https://github.com/user-attachments/assets/85d99750-71cb-4fb7-b969-d671c9cc2d36)
10. Now that subscription is manual, run the "Send upcoming renewal notification" action again.
11. Because we set this to a plain text email, confirm the new subscription item table is correct in plain text 🙈👍 
![image](https://github.com/user-attachments/assets/fec90f36-3625-42e8-bc00-25e731b25414)

--

### 2. Testing the new hooks

This PR introduces new `woocommerce_subscriptions_email_before_subscription_info` and `woocommerce_subscriptions_email_after_subscription_info` hooks, as well as adds the subscriptions item table which is attached to the `woocommerce_subscriptions_email_order_details` hook.

For testing customisation options, you can use the following snippet:

```php
add_action( 'woocommerce_before_template_part', function( $template_name, $template_path, $located, $args ) {
	if ( in_array( $template_name, [ 'emails/subscription-info.php', 'emails/plain/subscription-info.php' ], true ) ) {
		echo '<p style="color:red;"><strong>[[ TESTING ]] Before subscription info template</strong></p>';
	}
}, 10, 4 );

add_action( 'woocommerce_after_template_part', function( $template_name, $template_path, $located, $args ) {
	if ( in_array( $template_name, [ 'emails/subscription-info.php', 'emails/plain/subscription-info.php' ], true ) ) {
		echo '<p style="color:red;"><strong>[[ TESTING ]] After subscription info template</strong></p>';
	}
}, 10, 4 );

add_action( 'woocommerce_subscriptions_email_order_details', function() {
	echo '<p style="color:red;"><strong>[[ TESTING ]] After subscription item table</strong></p>';
}, 20, 0 );
```

Rather than going through and running the WP Admin actions, I used the email preview to confirm this is working as intended.

1. Add the above snippet to your store.
2. Go to **WooCommerce > Settings > Emails** and scroll to the bottom of the page.
3. Using the dropdown, preview a Customer Notification email and confirm your customisations are there:
![image](https://github.com/user-attachments/assets/191b29ef-7cc2-4abf-8d87-4d73d1e5015e)

## Product impact
<!-- What products will this PR ship in? -->

- [x] Added changelog entry (or does not apply)

> [!IMPORTANT]
> Once we ship these changes, we'll need to update [our documentation](https://woocommerce.com/document/subscriptions/subscriptions-notifications/) to include the latest images.

Commit:
Update subscription-info template to include hooks before/after subscription table

Commit:
Remove author tags

Commit:
Add subscription item details to our customer notification emails

Commit:
Fix undefined plain_text

Commit:
Remove the 'Here are the details' which felt unecessary

Commit:
Bump version of template to 7.2.0 and set missing x.x.x to include the version it was introduced in

Commit:
Always pass  var to template and attach all vars to new hook

Commit:
Add changelog entry

Commit:
Small capitalisation fix

Commit:
Merge branch 'trunk' into fix/752-subscription-info-template

barryhughes commented on templates/emails/customer-notification-auto-trial-ending.php:
```diff
@@ -59,18 +59,13 @@
 // Show subscription details.
 \WC_Subscriptions_Email::subscription_details( $subscription, $order, $sent_to_admin, $plain_text, true );
 
+do_action( 'woocommerce_subscriptions_email_order_details', $subscription, $sent_to_admin, $plain_text, $email );
```
Something of a passing thought rather than a request for a change here in this PR:

- So I don't think we have a requirement for documenting hooks in this project, and in any case this is admittedly pretty clear and readable (to me) without.
- Yet, if I was authoring an action callback for `woocommerce_subscriptions_email_order_details` would I expect the first param to be of type `WC_Subscription`, as this implies, or of type `WC_Order`, as [this implies](https://github.com/Automattic/woocommerce-subscriptions-core/blob/7.9.0/templates/emails/admin-new-renewal-order.php#L23)?
- One is of course a subclass of the other, but it still feels like there is the potential for subtle bugs (or at least friction when developing on top of hooks like this, where the ostensible type may vary—but the 3PD may not initially recognize that), and so a param comment *could* make the situation (slightly) clearer.

Again, mostly talking out loud rather than requesting a change here.

mattallan replied:
Thanks @barryhughes for highlighting this and I agree!

I've pushed aa738eab to add docblocks to these hooks. Let me know if it addresses the concerns :)

barryhughes replied:
That's great, thank you!

One last thought would be that we could document just one of the relevant hooks, and then 'point' to that from the others (as described in [4.1 Duplicate Hooks](https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#4-hooks-actions-and-filters)), something like this:

```php
/** This action is documented in templates/emails/file-with-canonical-docs.php */
```

The advantage being that, if we wanted to amend the docblock later, there is only one hook to update and therefore less chance of any single docblock being 'out of sync' with the others ... *but* this is perhaps getting nitpicky, and I'm not sure that even WP Core follows this all of the time.

mattallan replied:
Not nitpicky at all and I'm glad you suggested this! You had me sold on amending one docblock instead of 10.

Addressed in bfb21f5f59056568b8b53796fc7260342b7e785b

barryhughes commented on templates/emails/plain/subscription-info.php:
```diff
@@ -17,6 +16,8 @@
 $has_automatic_renewal = false;
 $is_parent_order       = wcs_order_contains_subscription( $order, 'parent' );
 
+do_action( 'woocommerce_subscriptions_email_before_subscription_info', $subscriptions, $order, $has_automatic_renewal, $is_parent_order, $is_admin_email, $plain_text );
```
Another hook musing 🙃 

This same hook has 6 params here, but has 7 in the matching HTML template. This is definitely workable (callback authors can state their hook accepts 7 args and the 7th can be optional—it will then work in both cases), *but,* is this good dev-ex?

Could it introduce friction/frustration if they notice the 7-arg hook first, then find it isn't being invoked in a plain-text context (or vice versa) ... might it be crisper and clearer to include a 7th arg here (set to false or null), for instance?

mattallan replied:
Thanks for noticing this! I don't believe this was intended and think I may have added the 7th parameter later and forgot to add it to the plain text email templates.

As per my comment on https://github.com/Automattic/woocommerce-subscriptions-core/pull/770#discussion_r1931496461, I'm going to remove these new before/after hooks since developers can already use `woocommerce_before|after_template_part` to achieve the same thing.

mattallan replied:
Thanks @barryhughes for noticing this !

As per https://github.com/Automattic/woocommerce-subscriptions-core/pull/770#discussion_r1931496461 I'm going to remove these additional do_actions in favour of having 3rd-parties to use the existing before/after hooks inside `wc_get_template()`.

We can look at introducing these again later if we're a proper need for them :)

Fixed in 661a67bc

barryhughes commented on templates/emails/plain/subscription-info.php:
```diff
@@ -61,3 +62,5 @@
 	);
 }
+
+do_action( 'woocommerce_subscriptions_email_after_subscription_info', $subscriptions, $order, $has_automatic_renewal, $is_parent_order, $is_admin_email, $plain_text );
```
🖐🏼 Question for my own education: we currently have a mix of hook prefixes in use (some like this use `woocommerce_subscriptions_*` and others use `wcs_*`) ... is this longer form what we're using going forward, or is it context-dependent?

barryhughes replied:
Thinking here that hooks represent a commitment, and also seeing that you mentioned the attractiveness of a wider refactoring, do we have a strong enough case here for introducing these new before/after hooks (considering a developer can use the existing `woocommerce_before_template_part` and `woocommerce_after_template_part` hooks)?

I'm not opposed, to be clear, and understand that specific hooks like this are superior in some respects, but still wanted to pose the question.

mattallan replied:
> Question for my own education: we currently have a mix of hook prefixes in use (some like this use woocommerce_subscriptions_* and others use wcs_*) ... is this longer form what we're using going forward, or is it context-dependent?

Good question!
I don't know when this started, but our preference changed to use the long-form "woocommerce_subscriptions_" to avoid any potential confusion and conflicts between similar features coming to WooCommerce like "WooCommerce Shipping" may also use the `wcs_` prefix.

It's a bit of a weird situation because all of our helper functions use `wcs_x_y_z()` and so changing to use `woocommerce_subscriptions_x_y_z()` prefix is going to result in very long function names, and I'm not really sure what the best alternative is.

We're pretty bad at doing this consistently but a general rule I try to stick to is:
- Use the `wcs_` prefix for functions inside a `x-functions.php` file
- Use `WC_Subscription_` for class names
- Use `woocommerce_subscriptions_` prefix for hooks

---

> considering a developer can use the existing `woocommerce_before_template_part` and `woocommerce_after_template_part` hooks

That's a really good point.

My preference would be to not create an extra level of hooks if there's already an alternative and so I'm going to just remove this.

barryhughes replied:
> ...but a general rule I try to stick to is:

Thank you, that's helpful! 👍🏼

barryhughes reviewed (approved):
Great! Tests well: thanks, @mattallan.

I left some questions and observations, some are just for my own benefit and none are blockers, so I'm hitting the *approve* button.

Commit:
Add missing 'skip_my_account_link' arg in plain text templates

Commit:
Remove additional hooks in favour of using woocommerce_before/after_template_part hooks

Commit:
Update changelog

Commit:
Add docblock to woocommerce_subscriptions_email_order_details hook.

mattallan commented:
Thanks @barryhughes for the review!

Summary of latest changes:
- [`661a67b`](https://github.com/Automattic/woocommerce-subscriptions-core/pull/770/commits/661a67bcf6bf67e40aa6aeb5c2a0cd26a87055cb) I've removed the new before/after hooks in favour of being able to use the existing `woocommerce_before|after_template_part` hooks (I've updated the code snippet in the PR description)
- [`aa738ea`](https://github.com/Automattic/woocommerce-subscriptions-core/pull/770/commits/aa738eab5bd63e847c4430f982a66219c9e2f650) - Adds documentation to the new do_actions added to all customer notification emails.

Would you mind taking another look at this? Thanks

barryhughes reviewed (approved):
Nice, thank you!

🪝 Left one more thought about hooks, however, definitely not any sort of blocker and can be ignored.

🔬 Re-tested, but using the updated instructions/snippets. Works great!

Commit:
Remove duplicate hook docblocks

mattallan commented:
Thanks @barryhughes for the review and testing. I've addressed the latest comment in bfb21f5f59056568b8b53796fc7260342b7e785b and will merge this once the checks turn green :)
